End of the .data removal in torch/optim

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34211 [WIP] End of the .data removal in torch/optim**

Follow up of https://github.com/pytorch/pytorch/pull/33640

The release note should mention the following:
This has the side effect of fixing inplace checks for the changes made by the optimizer.
There are two reasons for your code to start failing:
- The value of that parameter was actually saved and used and we were computing wrong gradients: You should replace code that uses `self.my_param` by `self.my_param.clone()` to make sure the saved version is different from the one that is modified by the optimizer.
- You know what you're doing and change the values back to the right thing before the next backward but the version counter cannot be decremented? Open an issue with your particular use case and we will help you to work around the version counter issue.

Differential Revision: [D20248684](https://our.internmc.facebook.com/intern/diff/D20248684)

albanD committed:
End of the .data removal in torch/optim

[ghstack-poisoned]

albanD committed:
Update on "End of the .data removal in torch/optim"


Differential Revision: [D20248684](https://our.internmc.facebook.com/intern/diff/D20248684)

[ghstack-poisoned]

vincentqb reviewed (changes requested):
As discussed, let's explicitly wait on the results from the internal tests.

vincentqb commented:
can you link the previous `.data` pr?

vincentqb reviewed (approved):
Internal tests have passed. LGTM.

albanD committed:
Update on "End of the .data removal in torch/optim"


Follow up of https://github.com/pytorch/pytorch/pull/33640

The release note should mention the following:
This has the side effect of fixing inplace checks for the changes made by the optimizer.
There are two reasons for your code to start failing:
- The value of that parameter was actually saved and used and we were computing wrong gradients: You should replace code that uses `self.my_param` by `self.my_param.clone()` to make sure the saved version is different from the one that is modified by the optimizer.
- You know what you're doing and change the values back to the right thing before the next backward but the version counter cannot be decremented? Open an issue with your particular use case and we will help you to work around the version counter issue.

Differential Revision: [D20248684](https://our.internmc.facebook.com/intern/diff/D20248684)

[ghstack-poisoned]

vincentqb commented on test/jit/test_data_parallel.py:
```diff
@@ -113,6 +113,9 @@ def test_tensor_sharing(self):
         second_forward = module.forward(first_forward)
 
+        # The replica must be re-created after the weights are changed
+        replica = dp.replicate(module, {0, 1})
```
As discussed with @albanD, this is not the correct way of modifying this test.

vincentqb reviewed (changes requested):
As discussed offline with @albanD, we need
- [x] a new test that does fail without the `.data` removal, but passes with it.
- [x] sign-off from someone in jit (e.g. @suo) that we can modify this test, so that it passes with and without `.data` removal
- [x] measure impact of replacing `module.forward(first_forward)` by `module.forward(x)` on line 114 in this test
- [x] all tests on CI external/internal to pass

albanD committed:
Update on "End of the .data removal in torch/optim"


Follow up of https://github.com/pytorch/pytorch/pull/33640

The release note should mention the following:
This has the side effect of fixing inplace checks for the changes made by the optimizer.
There are two reasons for your code to start failing:
- The value of that parameter was actually saved and used and we were computing wrong gradients: You should replace code that uses `self.my_param` by `self.my_param.clone()` to make sure the saved version is different from the one that is modified by the optimizer.
- You know what you're doing and change the values back to the right thing before the next backward but the version counter cannot be decremented? Open an issue with your particular use case and we will help you to work around the version counter issue.

Differential Revision: [D20248684](https://our.internmc.facebook.com/intern/diff/D20248684)

[ghstack-poisoned]

albanD committed:
Update on "[WIP] End of the .data removal in torch/optim"


Follow up of https://github.com/pytorch/pytorch/pull/33640

The release note should mention the following:
This has the side effect of fixing inplace checks for the changes made by the optimizer.
There are two reasons for your code to start failing:
- The value of that parameter was actually saved and used and we were computing wrong gradients: You should replace code that uses `self.my_param` by `self.my_param.clone()` to make sure the saved version is different from the one that is modified by the optimizer.
- You know what you're doing and change the values back to the right thing before the next backward but the version counter cannot be decremented? Open an issue with your particular use case and we will help you to work around the version counter issue.

Differential Revision: [D20248684](https://our.internmc.facebook.com/intern/diff/D20248684)

[ghstack-poisoned]

vincentqb commented:
Just to clarify:
* `test_tensor_sharing` is the new test: it fails without this patch, and succeeds with?
* `test_tensor_sharing_with_forward` is an updated version of the old test, where `module.forward(first_forward)` is replaced by `module(x)`?

There is still a handful of tests running on internal as of this moment :)

albanD commented:
> test_tensor_sharing is the new test: it fails without this patch, and succeeds with?

Not it is my version of making sure that the parameters and buffers are properly shared. It works both with and without the patch. And it has the advantage of not relying on `.data`.

> test_tensor_sharing_with_forward:

yes updated version with `x` and without using the `torch/optim` package as we cannot use it anymore if we want to keep the old logic.

vincentqb commented:
> yes updated version with `x` and without using the `torch/optim` package as we cannot use it anymore if we want to keep the old logic.

Alright, so that test fails without the patch, but not with. Great :)

vincentqb reviewed (approved):
Based on the discussion, I'm assuming the PR is no longer [WIP]. LGTM.
